### PR TITLE
[XLA:Python] Add CudaArrayInterfaceToBuffer

### PR DESCRIPTION
**Pull request by pearu**
As in the title.

Required by https://github.com/google/jax/pull/19233

### CODE REVIEW

**Comment by kamaljeeti** [comment]
Hi @pearu, can you sign up on this CLA provided below.
https://cla.developers.google.com/clas

**Comment by pearu** [comment]
@hawkinsp please review.

File: xla/python/py_array.cc [hunk 1]
```diff
@@ -701,6 +704,99 @@ py::dict PyArray::CudaArrayInterface() {
   return result;
 }
 
+StatusOr<pybind11::object> CudaArrayInterfaceToBuffer(const pybind11::dict& cai, std::shared_ptr<PyClient> client) {
+#ifndef GOOGLE_CUDA
+    throw XlaRuntimeError("This operation requires CUDA support.");
+#else
+  if (!cai.contains("data")) {
+    return InvalidArgument("CUDA Array Interface does not define `data`");
```
Comment (penpornk) [inline xla/python/py_array.cc:712]: Could you please use `absl::InvalidArgumentError` instead? We are trying to standardizing the calls. See below for examples
https://github.com/search?q=repo%3Aopenxla%2Fxla%20%22absl%3A%3Ainvalidargumenterror%22&type=code

Same for other places.

File: xla/python/py_array.cc [hunk 2]
```diff
@@ -701,6 +704,99 @@ py::dict PyArray::CudaArrayInterface() {
...
+  auto version = py::cast<int>(cai["version"]);
+  if (version < 2 || version > 3) {
+    LOG(WARNING) << "CUDA Array Interface version " << version << " support is undefined";
+  }
+  auto data = py::cast<pybind11::tuple>(cai["data"]);
+  auto data_value = pybind11::cast<std::intptr_t>(data[0]);
+  void* data_ptr = reinterpret_cast<void*>(data_value);
+  auto dimensions = pybind11::cast<std::vector<int64_t>>(cai["shape"]);
+  if (data_value == 0 && absl::c_find(dimensions, 0) == dimensions.end()) {
+    return InvalidArgument("CUDA Array Interface `data`(=NULL) and `shape`(no zero-valued dimensions) are inconsistent");
```
Comment (penpornk) [inline xla/python/py_array.cc:732]: This line is too long. Could you please apply a formatter? Same for other places.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR! I have quick formatting comments.

**Comment by kamaljeeti** [comment]
Hi @pearu , can you have a look at the comments? Thanks.

File: xla/python/py_array.cc (hunk 1 above)
Comment (pearu) [reply xla/python/py_array.cc:712]: Sure, thanks for the hint!
Fixed.

File: xla/python/py_array.cc (hunk 2 above)
Comment (pearu) [reply xla/python/py_array.cc]: Done. I hope using clang-format is OK.

**Comment by pearu** [comment]
@penpornk, this PR is ready for review.

**Review by penpornk — APPROVED** [review]
Sorry for the delay and thank you again for the PR!

**Comment by jvstokes** [comment]
Thank you for the PR, @pearu ! Could you also please update the python xla extension stub file to include cuda_array_interface_to_buffer, i.e. add it around here https://github.com/openxla/xla/blob/main/xla/python/xla_extension/__init__.pyi#L688

**Review by penpornk — APPROVED** [review]
Thank you for the update!

**Comment by pearu** [comment]
@penpornk could you check why Google internal checks failed?

**Comment by penpornk** [comment]
@pearu It seems the submit was blocked because of some flaky internal tests. I've triggered the submission again now.

**Comment by pearu** [comment]
@penpornk thanks for checking! Notice that there is a conflict, let me fix that first.

**Review by penpornk — APPROVED** [review]
Thank you for resolving the conflict!

**Pull request closed without merging**